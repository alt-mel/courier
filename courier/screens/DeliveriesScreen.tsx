import { useState, useEffect } from 'react';
import * as React from 'react';
import {
  View,
  Text,
  StyleSheet,
  Alert,
  FlatList,
  Pressable,
  TouchableWithoutFeedback,
  SafeAreaView
} from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { useQuery, gql } from '@apollo/client';

import { View as StyledView, Text as StyledText } from '../components/Themed';

export const GET_DELIVERIES_QUERY = gql`
  query myDeliveries {
    myDeliveries {
      id
      title
      price
      pickup_location
      destination_location
      description
      size
      weight
      status
    }
  }
`;

export default function DeliveriesScreen({ navigation }) {
  const [deliveries, setDeliveries] = useState([]);

  const { data, error, loading } = useQuery(GET_DELIVERIES_QUERY);

  useEffect(() => {
    if (error) {
      Alert.alert('Error fetching delivery', error.message);
    }
  }, [error]);

  useEffect(() => {
    if (loading) {
      console.log('Loading delivery');
    }
  }, [error]);

  useEffect(() => {
    if (data) {
      setDeliveries(data.myDeliveries);
    }
  }, [data]);

  if (!deliveries) {
    return null;
  }

  const onSubmit = async () => {
    try {
      await AsyncStorage.removeItem('token');
      navigation.navigate('Auth');
    } catch (exception) {
      return false;
    }
  };

  const selectItem = (item) => {
    console.log('Selected Item :', item);
    navigation.navigate('Track', {
      item: item
    });
  };

  const getHeader = () => {
    return (
      <Text style={styles.title} testID="Deliveries.Title">
        My Deliveries
      </Text>
    );
  };

  const getFooter = () => {
    return (
      <Pressable style={styles.button} onPress={onSubmit} disabled={loading}>
        <Text style={styles.buttonText}>Sign Out</Text>
      </Pressable>
    );
  };

  return (
    <SafeAreaView style={styles.container}>
      <FlatList
        data={deliveries}
        numColumns={1}
        renderItem={({ item }) => (
          <TouchableWithoutFeedback onPress={() => selectItem(item)}>
            <View style={styles.item} testID={`Deliveries.Item-${item.id}`}>
              <Text>{item.title}</Text>
              <Text>{item.status}</Text>
            </View>
          </TouchableWithoutFeedback>
        )}
        keyExtractor={(item) => item.id}
        ListHeaderComponent={getHeader}
        ListFooterComponent={getFooter}
      />
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginVertical: 10,
    marginHorizontal: 20
  },
  button: {
    alignItems: 'center',
    justifyContent: 'center',
    paddingVertical: 12,
    paddingHorizontal: 32,
    borderRadius: 4,
    elevation: 3,
    backgroundColor: '#bebebe'
  },
  buttonText: {
    color: 'white',
    fontSize: 18,
    fontWeight: 'bold'
  },
  title: {
    width: '100%',
    alignItems: 'center',
    fontSize: 24
  },
  item: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingTop: 15
  }
  // itemList: {
  //   margin: 8
  // }
});
