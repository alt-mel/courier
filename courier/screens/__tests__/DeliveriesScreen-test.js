import * as React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import { render, act } from '@testing-library/react-native';
import wait from 'waait';

import DeliveriesScreen, { GET_DELIVERIES_QUERY } from '../DeliveriesScreen';

const mocks = [
  {
    request: {
      query: GET_DELIVERIES_QUERY
    },
    result: {
      data: {
        myDeliveries: [
          {
            id: '61158567c849aff608e06253',
            title: 'box',
            price: '300',
            pickup_location: 'south.trombone.ask',
            destination_location: 'kiss.boo.run',
            description: 'Sample description for a package',
            size: 'Small',
            weight: '20 kilos',
            status: 'delivered'
          }
        ]
      }
    }
  }
];

it('renders the Deliveries Screen', () => {
  const tree = render(
    <MockedProvider>
      <DeliveriesScreen />
    </MockedProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders default elements', async () => {
  const { getByTestId } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <DeliveriesScreen />
    </MockedProvider>
  );

  await act(async () => {
    await wait(0);
  });

  expect(getByTestId('Deliveries.Title'));
  expect(getByTestId('Deliveries.Item-61158567c849aff608e06253'));
});
