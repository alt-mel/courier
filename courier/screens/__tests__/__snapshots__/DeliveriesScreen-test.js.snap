// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Deliveries Screen 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "marginHorizontal": 20,
      "marginVertical": 10,
    }
  }
>
  <RCTScrollView
    ListFooterComponent={[Function]}
    ListHeaderComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <Text
          style={
            Object {
              "alignItems": "center",
              "fontSize": 24,
              "width": "100%",
            }
          }
          testID="Deliveries.Title"
        >
          My Deliveries
        </Text>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.12)",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0,
              "elevation": 0,
              "minWidth": 64,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": true,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
            testID="Delivery.Button"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      undefined,
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "rgba(0, 0, 0, 0.32)",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Sign Out
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
